onnxruntime/core/framework/sequential_executor.cc

```diff
@@ -122,11 +123,218 @@ static Status ReleaseNodeMLValues(ExecutionFrame& frame,
+  if (seq_exec_plan.NodeHasFence(node_index)) {
+    for (int output_index = 0; output_index < op_kernel_context.OutputCount(); ++output_index) {
+      Fence_t fence = op_kernel_context.OutputFence(output_index);
+      if (fence) {
+        fence->AfterUsedAsOutput(queue_id);
+      }
+    }
+  }
+
+  return Status::OK();
+}
+
 Status SequentialExecutor::Execute(const SessionState& session_state, const std::vector<int>& feed_mlvalue_idxs,
                                    const std::vector<OrtValue>& feeds, const std::vector<int>& fetch_mlvalue_idxs,
-                                   std::vector<OrtValue>& fetches,
-                                   const std::unordered_map<size_t, CustomAllocator>& fetch_allocators,
+                                   std::vector<OrtValue>& fetches, const std::unordered_map<size_t, IExecutor::CustomAllocator>& fetch_allocators,
                                    const logging::Logger& logger) {
+  const SequentialExecutionPlan& seq_exec_plan = *session_state.GetExecutionPlan();
+  const auto& exec_plan_vec = seq_exec_plan.execution_plan;
+  const auto exec_plan_size = exec_plan_vec.size();
+  ExecutionFrame frame{feed_mlvalue_idxs, feeds, fetch_mlvalue_idxs, fetches, fetch_allocators, session_state};
+  ORT_RETURN_IF_ERROR(Execute(session_state, feeds, fetch_mlvalue_idxs, logger, frame, 0, exec_plan_size));
+  ORT_RETURN_IF_ERROR(FetchOutputsFromExecutionFrame(frame, session_state.GetTransferIntermidiateTensorOwnership(), fetches, logger));
+  return Status::OK();
+}
+
+Status SequentialExecutor::ExecutePartial(const SessionState& session_state, const std::vector<int>& feed_mlvalue_idxs,
+                                          const std::vector<OrtValue>& feeds, const std::vector<int>& fetch_mlvalue_idxs,
+                                          std::vector<OrtValue>& fetches, const std::unordered_map<size_t, IExecutor::CustomAllocator>& fetch_allocators,
+                                          const logging::Logger& logger, int64_t run_id) {
+  const SequentialExecutionPlan& seq_exec_plan = *session_state.GetExecutionPlan();
+  const auto& exec_plan_vec = seq_exec_plan.execution_plan;
+  const auto exec_plan_size = exec_plan_vec.size();
+  std::shared_ptr<ExecutionFrame> frame = nullptr;
+  std::map<int64_t, std::pair<std::unique_ptr<onnxruntime::ExecutionFrame>, size_t>>::iterator it;
+  size_t program_counter = 0;
+  //REVIEW(codemzs): We can get rid of this const cast by basically making partial graph exec manager APIs as SessionState.
+  PartialGraphExecutionManager& partial_graph_runs_manager = const_cast<SessionState&>(session_state).GetPartialGraphExecutionManager();
+  if (partial_graph_runs_manager.IsToBeExecutedRunId(run_id)) {
+    frame = std::shared_ptr<ExecutionFrame>(new ExecutionFrame(feed_mlvalue_idxs, feeds, fetch_mlvalue_idxs, fetches, fetch_allocators, session_state));
+    partial_graph_runs_manager.PreparePartialRun(frame, run_id, program_counter);
```

Comment:
Is there a way to do this when `CreatePartialRun()` is called? That would simplify the state space for the `PartialGraphExecutionManager`.

Reply:
I believe this was your previous comment: "I'd suggest having an explicit CreatePartialRun() method to allocate a run_id, so that you can avoid the complexity associated with DEFAULT_PARTIAL_RUN_ID."

Yes, you are right if I were to combine the two in one then it would reduce the state space by half but it also sort of contradicts your previous comment ;-) ...I can consider doing this in the very last commit. 

---
In reply to: [600938591](https://github.com/microsoft/onnxruntime/pull/7077#discussion_r600938591) [](ancestors = 600938591)

Reply:
No, I think you misunderstand.

During `CreatePartialRun()`, you should also be able to do the work to "prepare" the partial run as well, before inserting it into the PartialGraphExecutionManager. This avoids ever inserting "unprepared" runs in the PartialGraphExecutionManager, and avoids needing a "DEFAULT_PARTIAL_RUN_ID".

Reply:
@mrry To "prepare" you need feeds and fetches, but `CreatePartialRun()` only takes int64_t as argument, so now we are going back to the initial API. Its ok, I can make this change, sounds fine to me. `DEFAULT_PARTIAL_RUN_ID` does not exist in the code anymore .... 

---
In reply to: [602594652](https://github.com/microsoft/onnxruntime/pull/7077#discussion_r602594652) [](ancestors = 602594652)